xarray/core/dtypes.py

```diff
@@ -263,19 +320,18 @@ def result_type(
     if xp is None:
         xp = get_array_namespace(arrays_and_dtypes)
 
-    types = {
-        array_api_compat.result_type(preprocess_types(t), xp=xp)
-        for t in arrays_and_dtypes
-    }
-    if any(isinstance(t, np.dtype) for t in types):
-        # only check if there's numpy dtypes – the array API does not
-        # define the types we're checking for
-        for left, right in PROMOTE_TO_OBJECT:
-            if any(np.issubdtype(t, left) for t in types) and any(
-                np.issubdtype(t, right) for t in types
-            ):
-                return np.dtype(object)
-
+    if should_promote_to_object(arrays_and_dtypes, xp):
+        return np.dtype(object)
     return array_api_compat.result_type(
-        *map(preprocess_types, arrays_and_dtypes), xp=xp
+        *map(
+            functools.partial(
+                maybe_promote_to_variable_width,
+                # let extension arrays handle their own str/bytes
+                should_return_str_or_bytes=any(
+                    map(utils.is_allowed_extension_array_dtype, arrays_and_dtypes)
+                ),
+            ),
+            arrays_and_dtypes,
+        ),
+        xp=xp,
     )
```

Comment:
can we make this a bit less complex, for example by assigning the `partial` call to a variable?

Something like
```python
maybe_promote = functools.partial(
    maybe_promote_to_variable_width,
    # let extension arrays handle their own str/bytes
    should_return_str_or_bytes=any(
        map(utils.is_allowed_extension_array_dtype, arrays_and_dtypes)
    ),
)
return array_api_compat.result_type(*map(maybe_promote, arrays_and_dtypes), xp=xp)
```